File: 백준/Gold/13023. ABCDE/ABCDE.py
import sys
sys.setrecursionlimit(10000)
N, M = map(int, input().split())
graph = [[] for _ in range(N + 1)]

for _ in range(M):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

visited = [False] * (N + 1)
arrive = False


def dfs(v, depth):
    global arrive
    if depth == 5:
        arrive = True
        return
    visited[v] = True
    if not arrive:
        for i in graph[v]:
            if not visited[i]:
                dfs(i, depth+1)
    visited[v] = False

for i in range(1, N+1):
    dfs(i,1)
    if arrive:
        break

if arrive:
    print(1)
else:
    print(0)

File: 백준/Gold/13023. ABCDE/ABCDE.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static int N, M;
    static List<Integer>[] graph;
    static boolean arrive;
    static boolean[] visited;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        graph = new ArrayList[N + 1];
        visited = new boolean[N + 1];
        arrive = false;
        for(int i = 0; i<=N; i++){
            graph[i] = new ArrayList<>();
        }
        for(int i = 0; i<M; i++){
            st = new StringTokenizer(br.readLine());
            int s = Integer.parseInt(st.nextToken());
            int e = Integer.parseInt(st.nextToken());
            graph[s].add(e);
            graph[e].add(s);
        }

        for(int i = 1; i<=N; i++){
            dfs(i, 1);
            if(arrive) break;
        }
        if(arrive) System.out.println("1");
        else System.out.println("0");
    }

    private static void dfs(int v, int depth) {
        if(depth == 5 || arrive){
            arrive = true;
            return;
        }
        visited[v] = true;
        for(int i : graph[v]){
            if(!visited[i]) dfs(i, depth+1);
        }
        visited[v] = false;
    }
}
